snuba/replacer.py

```diff
@@ -381,3 +403,64 @@ def flush_batch(self, batch: Sequence[Replacement]) -> None:
             logger.info(
                 "Optimized %s partitions on %s" % (num_dropped, clickhouse_read.host)
             )
+
+    def _message_already_processed(self, metadata: ReplacementMessageMetadata) -> bool:
+        """
+        Figure out whether or not the message was already processed.
+
+        Check if there exists a recorded offset that took too long to execute.
+        If there is, we can conclude whether or not to process the incoming
+        message by comparing the incoming message to the recorded offset.
+        """
+        # float("-inf") implies this is the first message this worker is processing since startup
+        if self.__prolonged_offset == float("-inf"):
+            key = self._build_topic_group_index_key(metadata)
+            processed_offset = redis_client.get(key)
+            try:
+                self.__prolonged_offset = (
+                    -1 if processed_offset is None else int(processed_offset)
+                )
+            except ValueError as e:
+                redis_client.delete(key)
+                logger.warning(
+                    "Unexpected value found for an offset in Redis", exc_info=e,
+                )
+
+        return metadata.offset <= self.__prolonged_offset
+
+    def _check_timing_and_write_to_redis(
+        self, replacement: Replacement, start_time: float
+    ) -> None:
+        """
+        Write the offset just processed to Redis if execution took longer than the threshold.
+        Also store the offset locally to avoid Read calls to Redis.
+
+        If the Consumer dies while the insert query for the message was being executed,
+        the most recently executed offset would be present in Redis.
+        """
+        if (time.time() - start_time) < settings.REPLACER_PROCESSING_TIMEOUT_THRESHOLD:
+            return
+        message_metadata = replacement.get_message_metadata()
+        key = self._build_topic_group_index_key(message_metadata)
+        redis_client.set(
+            key,
+            message_metadata.offset,
+            ex=settings.REPLACER_PROCESSING_TIMEOUT_THRESHOLD_KEY_TTL,
+        )
+        self.__prolonged_offset = message_metadata.offset
+
+    def _build_topic_group_index_key(
+        self, message_metadata: ReplacementMessageMetadata
+    ) -> str:
+        """
+        Builds a unique key for a message being processed for a specific
+        topic, consumer group, and partition.
+        """
+        return ":".join(
+            [
+                "processed",
+                message_metadata.topic_name,
```

Comment:
Why does the topic name have to be in the replacement metadata? The individual replacement does not need to know about topics and consumer groups.
Also I think you will have to add the consumer group to the cache key as well.

Reply:
The topic is lost after `process_message()` since it just returns a None or Replacement. It needs to be in the metadata for the replacement itself unless there's another way to save that context for the offset. Same goes for the partition.

I can store the topic and partition as variables local to the worker itself, but that assumes the worker will always be reading from the same topic+partition. I'm not sure if this is the case.

As for the consumer group, I didn't see a way to get that info from a Kafka message itself but I believe the replacer processor's state is enough to identify a difference in consumer groups, so I added that state to the cache key.

Reply:
You can get rid of topic from here. Replacements would always be written to the same topic. Partition assignments for a consumer could change.

Having consumer group in the key would be good since what you are tracking is consumer group's progression across partitions. So it seems logical to have that in the key.

You should be able to use consumer group from https://github.com/getsentry/snuba/blob/master/snuba/datasets/events_processor_base.py#L180-L182 and pass it to https://github.com/getsentry/snuba/blob/master/snuba/cli/replacer.py#L119

Reply:
Alright, I've got the consumer group and topic passed to the Worker on instantiation. ReplacerWorker now requires these params:
```python
storage: WritableTableStorage,
replacements_topic: str,
consumer_group: str,
metrics: MetricsBackend
```

The 2 new strings are just saved locally to the worker and are used when building the cache key.